File: src/main/java/ir/ac/sbu/redispark/ConnectionPool.java
package ir.ac.sbu.redispark;

import redis.clients.jedis.Jedis;
import redis.clients.jedis.JedisPool;
import redis.clients.jedis.JedisPoolConfig;
import redis.clients.jedis.Pipeline;

import java.io.Serializable;
/**
 *
 */
public class ConnectionPool implements Serializable {

    private JedisPool jedisPool;

    public ConnectionPool(String host, int port) {
            JedisPoolConfig poolConfig = new JedisPoolConfig();
            poolConfig.setMaxTotal(1000);
            poolConfig.setMaxIdle(1);
            poolConfig.setTestOnBorrow(true);
            poolConfig.setTestOnReturn(true);
            poolConfig.setTestWhileIdle(true);
            poolConfig.setMinEvictableIdleTimeMillis(60000);
            poolConfig.setTimeBetweenEvictionRunsMillis(3000);
            poolConfig.setNumTestsPerEvictionRun(3);
            jedisPool = new JedisPool(poolConfig, host, port, 60000);
    }

    public Jedis getJedis() {
        return jedisPool.getResource();
    }

    public void close() {
        jedisPool.close();
    }
}

File: src/main/scala/ir/ac/sbu/redispark/ConnectionPool.scala
package ir.ac.sbu.redispark

import java.util.concurrent.ConcurrentHashMap

import redis.clients.jedis.exceptions.JedisConnectionException
import redis.clients.jedis._

import scala.collection.JavaConversions._


object ConnectionPool {

    @transient private lazy val pools: ConcurrentHashMap[RedisEndpoint, ShardedJedisPool] =
        new ConcurrentHashMap[RedisEndpoint, ShardedJedisPool]()

    def connect(re: RedisEndpoint): ShardedJedis = {
        val pool = pools.getOrElseUpdate(re, {
            val shardInfo = List(new JedisShardInfo(re.host, re.port))

            val poolConfig: JedisPoolConfig = new JedisPoolConfig();
            poolConfig.setMaxTotal(10000)
            poolConfig.setMaxIdle(10)
            poolConfig.setTestOnBorrow(true)
            poolConfig.setTestOnReturn(true)
            poolConfig.setTestWhileIdle(true)
            poolConfig.setMinEvictableIdleTimeMillis(60000)
            poolConfig.setTimeBetweenEvictionRunsMillis(30000)
            poolConfig.setNumTestsPerEvictionRun(3)
            new ShardedJedisPool(poolConfig, shardInfo)
        }
        )
        var sleepTime: Int = 4
        var conn: ShardedJedis = null
        while (conn == null) {
            try {
                conn = pool.getResource
            }
            catch {
                case e: JedisConnectionException if e.getCause.toString.
                  contains("ERR max number of clients reached") => {
                    if (sleepTime < 500) sleepTime *= 2
                    Thread.sleep(sleepTime)
                }
                case e: Exception => throw e
            }
        }
        conn
    }
}

